src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,144 @@
+
+import se.liu.ida.hefquin.engine.data.VocabularyMapping;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpLocalToGlobal;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpMultiwayUnion;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.impl.poptimizer.utils.PhysicalPlanWithStatistics;
+
+import java.util.*;
+
+/**
+ * This class implements a query optimizer that builds left-deep query plans,
+ * for which it uses a greedy approach to determine the join order based on vocabulary mapping and cardinality estimation.
+ * Specifically, choose the first subPlan with the smallest cardinality, then the subPlan with the same vocabulary mapping is selected.
+ * If there are multiple subPlans with the same vocabulary mapping, the order depends on estimated cardinality;
+ * If such a subPlan does not exist, simply select the subPlan with the lowest cardinality as the next subPlan.
+ *
+ * After determining join ordering, the physical plan can be constructed using bind join.
+ */
+public class ReduceVocRewritingJoinPlanOptimizerImpl extends CardinalityBasedJoinPlanOptimizerBase
+{
+    public ReduceVocRewritingJoinPlanOptimizerImpl(final FederationAccessManager fedAccessMgr ) {
+        super(fedAccessMgr);
+    }
+
+    @Override
+    public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+
+        return new CardinalityBasedJoinPlanOptimizerBase.GreedyConstructionAlgorithm(subplans) {
+            @Override
+            protected PhysicalPlan determineJoinOrderAndConstructPlan( final PhysicalPlanWithStatistics firstSubPlan, final List<PhysicalPlanWithStatistics> subPlansWithStatistics, final Comparator<PhysicalPlanWithStatistics> orderBasedOnCard) {
+                final PriorityQueue<PhysicalPlanWithStatistics> orderedCandidateSubPlans = new PriorityQueue<>(orderBasedOnCard);
+                PhysicalPlanWithStatistics nextSelectedSubPlan = firstSubPlan;
+                PhysicalPlan currentJoinPlan = nextSelectedSubPlan.plan;
+
+                while ( !subPlansWithStatistics.isEmpty() ){
+                    // Identify subplans that have join variables with the selected subplan as new candidateSubPlans.
+                    List<PhysicalPlanWithStatistics> tmpCandidateSubPlans = getSubPlansContainVars( currentJoinPlan.getExpectedVariables(), subPlansWithStatistics );
+                    if( tmpCandidateSubPlans.isEmpty() ){
+                        tmpCandidateSubPlans = subPlansWithStatistics;
+                    }
+
+                    // Keep candidate SubPlans that use the same vocabulary mapping with the previous selected SubPlan.
```

Comment:
What if the previously selected subplan has multiple vocabulary mappings associated with it? Would that still be okay here?

Reply:
I suppose the previously selected sub-plan has multiple vocabulary mappings only when this sub-plan has union as the root operator. Given that these candidate plans will be utilized to determine the join order, selecting a sub-plan that has some common vocabulary mapping may increase the likelihood of creating a plan with tpAdd that facilitates a reduction in plan rewriting operations. What do you think?